### DISCUSSION TITLE
Replace non-performance-critical opcodes by call to methods of an internal object.

### DISCUSSION THREAD

**Discussion by iritkatriel** [faster-cpython/ideas#202] [Ideas]
There are a number of opcodes which are not performance critical and which seem to exist for the sole purpose of invoking a function or loading a non-picklable constant to the stack (which could be returned from a function).  Examples:

LOAD_ASSERTION_ERROR, IMPORT_NAME, IMPORT_FROM, PRINT_EXPR, PREP_RERAISE_STAR.

Perhaps we could create a simple (picklable) object that has these functions as its methods, so that the compiler can load this object to the stack and invoke the methods via a series of generic opcodes. This object should ideally not be accessible from python, it should be stateless, it certainly should not be subclassable and the methods should be immutable.

This is an alternative to an idea that came up previously, to load AssertionError via a new [LOAD_COMMON_CONST](https://github.com/faster-cpython/ideas/issues/65) opcode that fetches it from an internal consts array. That approach could be good for performance critical consts, but for the slow ones we don’t actually need an opcode at all.

**Comment by ericsnowcurrently** [comment]
Seems worth exploring.

**Comment by markshannon** [comment]
Other candidates for lowering are:

`LOAD_BUILD_CLASS` can be replaced with `LOAD_COMMON_CONST build_class`
`RAISE_VARARGS n` can be replaced with `LOAD_COMMON_CONST raise_varargs; PRECALL_FUNCTION; CALL n`
`BEFORE_ASYNC_WITH` and `BEFORE_WITH` can be replaced with the lower level operations they are composed of. These will need a new `LOAD_SPECIAL` instruction.

`BEFORE_WITH` becomes
```
     COPY 1
     LOAD_SPECIAL "__exit__"
     SWAP 2
     LOAD_SPECIAL "__enter__"
     PRECALL_FUNCTION 
     CALL 0
```

**Comment by markshannon** [comment]
If we add a `LOAD_LOCAL` instruction (which would do the same as `LOAD_NAME` without falling back to the globals), then we can lower a couple more instructions, as `try-except` is now cheap.

`LOAD_CLASSDEREF name` becomes
```
try:
     LOAD_LOCAL name
except KeyError:
     LOAD_DEREF name
```

`SETUP_ANNOTATION` becomes
```
try:
     LOAD_LOCAL __annotations__
except:
    BUILD_MAP
    STORE_NAME __annotations__
```

**Comment by gvanrossum** [comment]
How cheap is raising an exception? 99.9% of LOAD_CLASSDEREF will raise NameError, which means a NameError object is created and immediately discarded.

**Comment by markshannon** [comment]
Raising an exception is expensive. Relative to the simple lookup, very expensive. So let's not do that.
A LBYL approach would work better; adding `HAS_LOCAL` instead of `LOAD_LOCAL`.

`LOAD_CLASSDEREF name` then becomes:
```
    HAS_LOCAL name
    POP_JUMP_IF_FALSE use_deref
    LOAD_LOCAL name
    JUMP exit
use_deref:
    LOAD_DEREF name
exit:
```
and `SETUP_ANNOTATION` becomes:
```
    HAS_LOCAL __annotations__
    POP_JUMP_IF_TRUE done
    BUILD_MAP
    STORE_NAME __annotations__
done:
```

**Comment by markshannon** [comment]
https://github.com/python/cpython/issues/99005
